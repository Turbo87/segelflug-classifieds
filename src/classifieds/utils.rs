use std::collections::HashSet;

pub fn strip_html(value: &str) -> String {
    ammonia::Builder::new()
        .tags(HashSet::new())
        .clean(value)
        .to_string()
}
