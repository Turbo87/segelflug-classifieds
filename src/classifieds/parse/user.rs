use crate::classifieds::utils::strip_html;
use scraper::{Html, Selector};

#[derive(Debug)]
#[allow(dead_code)]
pub struct ClassifiedsUser {
    pub name: Option<String>,
    pub address: Option<String>,
    pub location: Option<String>,
    pub website: Option<String>,
}

impl From<&str> for ClassifiedsUser {
    #[instrument(name = "ClassifiedsUser::from", skip(text))]
    fn from(text: &str) -> Self {
        lazy_static! {
            static ref NAME_SELECTOR: Selector = Selector::parse("li.name").unwrap();
            static ref ADDRESS_SELECTOR: Selector = Selector::parse("li.address").unwrap();
            static ref LOCATION_SELECTOR: Selector = Selector::parse("li.location").unwrap();
            static ref WEBSITE_SELECTOR: Selector = Selector::parse("li.website").unwrap();
        }

        let html = Html::parse_document(text);
        if !html.errors.is_empty() {
            debug!("found HTML parsing errors: {:?}", html.errors);
        }

        let name = html
            .select(&NAME_SELECTOR)
            .next()
            .map(|element| element.inner_html())
            .map(|html| strip_html(&html))
            .map(|text| text.trim().to_string());

        let address = html
            .select(&ADDRESS_SELECTOR)
            .next()
            .map(|element| element.inner_html())
            .map(|html| strip_html(&html))
            .map(|text| text.replace("Adresse:", ""))
            .map(|text| text.trim().to_string());

        let location = html
            .select(&LOCATION_SELECTOR)
            .next()
            .map(|element| element.inner_html())
            .map(|html| strip_html(&html))
            .map(|text| text.replace("Standort:", ""))
            .map(|text| text.trim().to_string());

        let website = html
            .select(&WEBSITE_SELECTOR)
            .next()
            .map(|element| element.inner_html())
            .map(|html| strip_html(&html))
            .map(|text| text.trim().to_string());

        Self {
            name,
            address,
            location,
            website,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::ClassifiedsUser;
    use std::fs;

    #[test]
    fn parse_test() {
        glob!("test-input/user/*.html", |path| {
            let text = fs::read_to_string(path).unwrap();
            let user = ClassifiedsUser::from(text.as_str());
            assert_debug_snapshot!(user);
        });
    }
}
