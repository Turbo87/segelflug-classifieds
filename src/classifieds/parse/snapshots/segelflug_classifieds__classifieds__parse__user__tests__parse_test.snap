---
source: src/classifieds/parse/user.rs
expression: user
input_file: src/classifieds/parse/test-input/user/basic.html
snapshot_kind: text
---
ClassifiedsUser {
    name: Some(
        "Raatz",
    ),
    address: None,
    location: Some(
        "Germany",
    ),
    website: None,
}
